File: app/src/main/java/io/github/coffeegerm/materiallogbook/ui/EditEntryActivity.java
/*
 * Copyright 2018 Coffee and Cream Studios
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.coffeegerm.materiallogbook.ui;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.TimePicker;
import android.widget.Toast;

import java.util.Calendar;
import java.util.Date;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.github.coffeegerm.materiallogbook.MaterialLogbookApplication;
import io.github.coffeegerm.materiallogbook.R;
import io.github.coffeegerm.materiallogbook.data.model.EntryItem;
import io.github.coffeegerm.materiallogbook.utils.Utilities;
import io.realm.Realm;

import static io.github.coffeegerm.materiallogbook.utils.Constants.DATE_FORMAT;
import static io.github.coffeegerm.materiallogbook.utils.Constants.PREF_DARK_MODE;
import static io.github.coffeegerm.materiallogbook.utils.Constants.TWELVE_HOUR_TIME_FORMAT;

/**
 * Created by dyarz on 8/17/2017.
 * <p>
 * Activity created to allow user to edit their entries
 * already present in database
 */

public class EditEntryActivity extends AppCompatActivity {
  
  public static final String ITEM_ID = "itemId";
  
  @Inject
  public SharedPreferences sharedPreferences;
  @Inject
  public Utilities utilities;
  
  @BindView(R.id.toolbar)
  Toolbar toolbar;
  @BindView(R.id.cancel)
  Button cancel;
  @BindView(R.id.update)
  Button update;
  @BindView(R.id.new_entry_date)
  EditText date;
  @BindView(R.id.new_entry_time)
  EditText time;
  @BindView(R.id.new_entry_blood_glucose_level)
  EditText bloodGlucose;
  @BindView(R.id.new_entry_carbohydrates_amount)
  EditText carbohydrates;
  @BindView(R.id.new_entry_insulin_units)
  EditText insulin;
  @BindView(R.id.delete_entry)
  TextView deleteEntry;
  @BindView(R.id.breakfast_status)
  ImageButton breakfast;
  @BindView(R.id.lunch_status)
  ImageButton lunch;
  @BindView(R.id.dinner_status)
  ImageButton dinner;
  @BindView(R.id.sick_status)
  ImageButton sick;
  @BindView(R.id.exercise_status)
  ImageButton exercise;
  @BindView(R.id.sweets_status)
  ImageButton sweets;
  
  Handler handler;
  /* Original values from item. Compare to possible updated values to find what needs to be updated in database */
  Date originalDate;
  int originalStatus;
  int originalBloodGlucose;
  int originalCarbohydrates;
  double originalInsulin;
  /* items to be used to altered to show that the item has been updated */
  int updatedStatus;
  int updatedBloodGlucose;
  int updatedCarbohydrates;
  double updatedInsulin;
  private Calendar originalCalendar;
  private Calendar updatedCalendar;
  private Realm realm;
  private String itemIdString;
  private EntryItem item;
  
  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    MaterialLogbookApplication.syringe.inject(this);
    if (sharedPreferences.getBoolean(PREF_DARK_MODE, false))
      setTheme(R.style.AppTheme_Dark);
    setContentView(R.layout.activity_edit_entry);
    ButterKnife.bind(this);
    realm = Realm.getDefaultInstance();
    handler = new Handler();
    setSupportActionBar(toolbar);
    if (getSupportActionBar() != null)
      getSupportActionBar().setTitle(R.string.edit_entry_toolbar);
    itemIdString = getIntent().getStringExtra(ITEM_ID);
    item = getItem();
    updatedCalendar = Calendar.getInstance();
    getOriginalValues(); // must call before hints are set
    setHints();
    
    date.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        DatePickerDialog dialog = new DatePickerDialog(EditEntryActivity.this,
              (Build.VERSION.SDK_INT >= 21 ? android.R.style.Theme_Material_Dialog_Alert
                    : android.R.style.Theme_Holo_Light_Dialog),
              new DatePickerDialog.OnDateSetListener() {
                @Override
                public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                  month++;
                  date.setText(utilities.formatDate(month, dayOfMonth, year));
                  month--;
                  updatedCalendar.set(year, month, dayOfMonth);
                }
              }, originalCalendar.get(Calendar.YEAR), // year
              originalCalendar.get(Calendar.MONTH), // month
              originalCalendar.get(Calendar.DAY_OF_MONTH)); // day
        
        if (Build.VERSION.SDK_INT < 21)
          if (dialog.getWindow() != null)
            dialog.getWindow().setBackgroundDrawable(new ColorDrawable(0));
        
        dialog.show();
      }
    });
    
    time.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        TimePickerDialog timePickerDialog = new TimePickerDialog(EditEntryActivity.this,
              new TimePickerDialog.OnTimeSetListener() {
                @Override
                public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
                  time.setText(utilities.checkTimeString(hourOfDay, minute));
                  updatedCalendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
                  updatedCalendar.set(Calendar.MINUTE, minute);
                }
              },
              originalCalendar.get(Calendar.HOUR_OF_DAY), // current hour
              originalCalendar.get(Calendar.MINUTE), // current minute
              false); //no 24 hour view
        timePickerDialog.show();
      }
    });
    
    bloodGlucose.addTextChangedListener(new TextWatcher() {
      @Override
      public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
      
      }
      
      @Override
      public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
      
      }
      
      @Override
      public void afterTextChanged(Editable editable) {
        updatedBloodGlucose = Integer.parseInt(editable.toString());
      }
    });
    
    carbohydrates.addTextChangedListener(new TextWatcher() {
      @Override
      public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
      
      }
      
      @Override
      public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
      
      }
      
      @Override
      public void afterTextChanged(Editable editable) {
        updatedCarbohydrates = Integer.parseInt(editable.toString());
      }
    });
    
    insulin.addTextChangedListener(new TextWatcher() {
      @Override
      public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
      
      }
      
      @Override
      public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
      
      }
      
      @Override
      public void afterTextChanged(Editable editable) {
        updatedInsulin = Double.parseDouble(editable.toString());
      }
    });
    
    breakfast.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        updatedStatus = 1;
        setStatus(updatedStatus);
        final Toast toast = Toast.makeText(getApplicationContext(), "Breakfast", Toast.LENGTH_SHORT);
        toast.setGravity(Gravity.CENTER, 0, 300);
        toast.show();
        handler.postDelayed(new Runnable() {
          @Override
          public void run() {
            toast.cancel();
          }
        }, 700);
      }
    });
    
    lunch.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        updatedStatus = 2;
        setStatus(updatedStatus);
        final Toast toast = Toast.makeText(getApplicationContext(), "Lunch", Toast.LENGTH_SHORT);
        toast.setGravity(Gravity.CENTER, 0, 300);
        toast.show();
        handler.postDelayed(new Runnable() {
          @Override
          public void run() {
            toast.cancel();
          }
        }, 700);
      }
    });
    
    dinner.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        updatedStatus = 3;
        setStatus(updatedStatus);
        final Toast toast = Toast.makeText(getApplicationContext(), "Dinner", Toast.LENGTH_SHORT);
        toast.setGravity(Gravity.CENTER, 0, 300);
        toast.show();
        handler.postDelayed(new Runnable() {
          @Override
          public void run() {
            toast.cancel();
          }
        }, 700);
      }
    });
    
    sick.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        updatedStatus = 4;
        setStatus(updatedStatus);
        final Toast toast = Toast.makeText(getApplicationContext(), "Sick", Toast.LENGTH_SHORT);
        toast.setGravity(Gravity.CENTER, 0, 300);
        toast.show();
        handler.postDelayed(new Runnable() {
          @Override
          public void run() {
            toast.cancel();
          }
        }, 700);
      }
    });
    
    exercise.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        updatedStatus = 5;
        setStatus(updatedStatus);
        final Toast toast = Toast.makeText(getApplicationContext(), "Exercise", Toast.LENGTH_SHORT);
        toast.setGravity(Gravity.CENTER, 0, 300);
        toast.show();
        handler.postDelayed(new Runnable() {
          @Override
          public void run() {
            toast.cancel();
          }
        }, 700);
      }
    });
    
    sweets.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        updatedStatus = 6;
        setStatus(updatedStatus);
        final Toast toast = Toast.makeText(getApplicationContext(), "Sweets", Toast.LENGTH_SHORT);
        toast.setGravity(Gravity.CENTER, 0, 300);
        toast.show();
        handler.postDelayed(new Runnable() {
          @Override
          public void run() {
            toast.cancel();
          }
        }, 700);
      }
    });
    
    deleteEntry.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        final AlertDialog.Builder builder;
        
        // Sets theme based on VERSION_CODE
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
          builder = new AlertDialog.Builder(EditEntryActivity.this, android.R.style.Theme_Material_Dialog_NoActionBar);
        else builder = new AlertDialog.Builder(EditEntryActivity.this);
        
        builder.setTitle(R.string.delete_single_entry)
              .setMessage(R.string.delete_single_entry_message)
              .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int which) {
                  // continue with delete
                  realm.executeTransaction(new Realm.Transaction() {
                    @Override
                    public void execute(@NonNull Realm realm) {
                      item.deleteFromRealm();
                      Toast.makeText(EditEntryActivity.this, R.string.entry_deleted, Toast.LENGTH_SHORT).show();
                      finish();
                    }
                  });
                }
              })
              .setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int which) {
                  // do nothing
                  dialog.dismiss();
                }
              })
              .setIcon(R.drawable.ic_trash)
              .show();
      }
    });
    
    cancel.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        finish();
      }
    });
    
    update.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        updateEntry();
      }
    });
  }
  
  private EntryItem getItem() {
    return realm.where(EntryItem.class).equalTo("id", itemIdString).findAll().get(0);
  }
  
  private void getOriginalValues() {
    originalStatus = item.getStatus();
    originalCalendar = Calendar.getInstance();
    originalCalendar.setTime(item.getDate());
    originalDate = originalCalendar.getTime();
    originalBloodGlucose = item.getBloodGlucose();
    originalCarbohydrates = item.getCarbohydrates();
    originalInsulin = item.getInsulin();
  }
  
  private void setHints() {
    String formattedDate = DATE_FORMAT.format(originalDate);
    String formattedTime = TWELVE_HOUR_TIME_FORMAT.format(originalDate);
    date.setHint(formattedDate);
    time.setHint(formattedTime);
    bloodGlucose.setHint(String.valueOf(originalBloodGlucose));
    if (originalCarbohydrates != 0)
      carbohydrates.setHint(String.valueOf(originalCarbohydrates));
    else carbohydrates.setHint(R.string.dash);
    if (originalInsulin != 0.0)
      insulin.setHint(String.valueOf(originalInsulin));
    else insulin.setHint(R.string.dash);
    setStatus(originalStatus);
  }
  
  private void setStatus(int status) {
    Drawable status_checked = getResources().getDrawable(R.drawable.status_checked);
    Drawable status_unchecked = getResources().getDrawable(R.drawable.status_unchecked);
    switch (status) {
      case 1:
        breakfast.setBackground(status_checked);
        lunch.setBackground(status_unchecked);
        dinner.setBackground(status_unchecked);
        sick.setBackground(status_unchecked);
        exercise.setBackground(status_unchecked);
        sweets.setBackground(status_unchecked);
        break;
      case 2:
        breakfast.setBackground(status_unchecked);
        lunch.setBackground(status_checked);
        dinner.setBackground(status_unchecked);
        sick.setBackground(status_unchecked);
        exercise.setBackground(status_unchecked);
        sweets.setBackground(status_unchecked);
        break;
      case 3:
        breakfast.setBackground(status_unchecked);
        lunch.setBackground(status_unchecked);
        dinner.setBackground(status_checked);
        sick.setBackground(status_unchecked);
        exercise.setBackground(status_unchecked);
        sweets.setBackground(status_unchecked);
        break;
      case 4:
        breakfast.setBackground(status_unchecked);
        lunch.setBackground(status_unchecked);
        dinner.setBackground(status_unchecked);
        sick.setBackground(status_checked);
        exercise.setBackground(status_unchecked);
        sweets.setBackground(status_unchecked);
        break;
      case 5:
        breakfast.setBackground(status_unchecked);
        lunch.setBackground(status_unchecked);
        dinner.setBackground(status_unchecked);
        sick.setBackground(status_unchecked);
        exercise.setBackground(status_checked);
        sweets.setBackground(status_unchecked);
        break;
      case 6:
        breakfast.setBackground(status_unchecked);
        lunch.setBackground(status_unchecked);
        dinner.setBackground(status_unchecked);
        sick.setBackground(status_unchecked);
        exercise.setBackground(status_unchecked);
        sweets.setBackground(status_checked);
        break;
      default:
        break;
    }
  }
  
  private void updateEntry() {
    try {
      Date dateToSave = new Date();
      realm.beginTransaction();
      item.deleteFromRealm();
      EntryItem itemToSave = new EntryItem();
      itemToSave.setStatus(updatedStatus);
      if (updatedCalendar.getTimeInMillis() != originalCalendar.getTimeInMillis()) {
        dateToSave.setTime(updatedCalendar.getTimeInMillis());
        itemToSave.setDate(dateToSave);
      } else {
        dateToSave.setTime(originalCalendar.getTimeInMillis());
        itemToSave.setDate(dateToSave);
      }
      if (!bloodGlucose.getText().toString().equals("")) {
        itemToSave.setBloodGlucose(Integer.parseInt(bloodGlucose.getText().toString()));
      }
      if (!carbohydrates.getText().toString().equals("")) {
        itemToSave.setCarbohydrates(Integer.parseInt(carbohydrates.getText().toString()));
      }
      if (!insulin.getText().toString().equals("")) {
        itemToSave.setInsulin(Double.parseDouble(insulin.getText().toString()));
      }
      realm.copyToRealm(itemToSave);
      realm.commitTransaction();
    } finally {
      realm.close();
      finish();
    }
  }
  
  @Override
  public void onDestroy() {
    realm.close();
    super.onDestroy();
  }
}

File: app/src/main/java/io/github/coffeegerm/glucoseguide/ui/entry/EditEntryActivity.kt
/*
 * Copyright 2018 Coffee and Cream Studios
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.coffeegerm.glucoseguide.ui.entry

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Intent
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import io.github.coffeegerm.glucoseguide.GlucoseGuide
import io.github.coffeegerm.glucoseguide.R
import io.github.coffeegerm.glucoseguide.data.DatabaseManager
import io.github.coffeegerm.glucoseguide.data.model.Entry
import io.github.coffeegerm.glucoseguide.ui.MainActivity
import io.github.coffeegerm.glucoseguide.utils.Constants
import io.github.coffeegerm.glucoseguide.utils.DateFormatter
import io.github.coffeegerm.glucoseguide.utils.SharedPreferencesManager
import kotlinx.android.synthetic.main.activity_edit_entry.*
import java.util.*
import javax.inject.Inject


class EditEntryActivity : AppCompatActivity(), AdapterView.OnItemSelectedListener {
  
  companion object {
    var wasDateChanged = false
  }
  
  @Inject
  lateinit var sharedPreferencesManager: SharedPreferencesManager
  @Inject
  lateinit var databaseManager: DatabaseManager
  @Inject
  lateinit var dateFormatter: DateFormatter
  
  // Original values from old. Compare to possible updated values to find what needs to be updated in database
  private lateinit var originalDate: Date
  private var originalStatus: Int = 0
  private var originalBloodGlucose: Int = 0
  private var originalCarbohydrates: Int = 0
  private var originalInsulin: Double = 0.0
  
  // items to be used to altered to show that the old has been updated
  private var updatedStatus: Int = 0
  internal var updatedBloodGlucose: Int = 0
  internal var updatedCarbohydrates: Int = 0
  internal var updatedInsulin: Double = 0.0
  private var originalCalendar: Calendar = Calendar.getInstance()
  private var updatedCalendar: Calendar = Calendar.getInstance()
  
  private lateinit var itemId: String
  private lateinit var old: Entry
  
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    GlucoseGuide.syringe.inject(this)
    if (sharedPreferencesManager.getBoolean(Constants.PREF_DARK_MODE))
      setTheme(R.style.AppTheme_Dark)
    setContentView(R.layout.activity_edit_entry)
    setSupportActionBar(edit_entry_toolbar)
    supportActionBar?.setTitle(R.string.edit_entry_toolbar)
    supportActionBar?.setDisplayHomeAsUpEnabled(true)
    supportActionBar?.setDisplayShowHomeEnabled(true)
    
    itemId = intent.getStringExtra(Constants.ITEM_ID)
    old = databaseManager.getEntryFromId(itemId)!!
    
    getOriginalValues()
    
    val spinnerAdapter = ArrayAdapter<String>(this, android.R.layout.simple_spinner_dropdown_item, resources.getStringArray(R.array.status_selection))
    edit_status_selector.adapter = spinnerAdapter
    edit_status_selector.onItemSelectedListener = this
    
    setHints()
    
    edit_entry_date.setOnClickListener {
      val dialog = DatePickerDialog(this,
            DatePickerDialog.OnDateSetListener { _, year, month, dayOfMonth ->
              var correctMonth = month
              correctMonth++
              edit_entry_date.setText(dateFormatter.formatDateForEditText(correctMonth, dayOfMonth, year))
              correctMonth--
              if (!wasDateChanged) wasDateChanged = true
              updatedCalendar.set(year, correctMonth, dayOfMonth)
            }, originalCalendar.get(Calendar.YEAR),
            originalCalendar.get(Calendar.MONTH),
            originalCalendar.get(Calendar.DAY_OF_MONTH))
      
      dialog.show()
    }
    
    edit_entry_time.setOnClickListener {
      val timePickerDialog = TimePickerDialog(this,
            TimePickerDialog.OnTimeSetListener { _, hourOfDay, minute ->
              edit_entry_time.setText(dateFormatter.formatTimeForEditText(hourOfDay, minute))
              updatedCalendar.set(Calendar.HOUR_OF_DAY, hourOfDay)
              updatedCalendar.set(Calendar.MINUTE, minute)
              if (!wasDateChanged) wasDateChanged = true
            },
            originalCalendar.get(Calendar.HOUR_OF_DAY),
            originalCalendar.get(Calendar.MINUTE),
            false)
      timePickerDialog.show()
    }
    
    edit_entry_blood_glucose_level.addTextChangedListener(object : TextWatcher {
      override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
      
      }
      
      override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
      
      }
      
      override fun afterTextChanged(editable: Editable) {
        updatedBloodGlucose = Integer.parseInt(editable.toString())
      }
    })
    
    edit_entry_carbohydrates_amount.addTextChangedListener(object : TextWatcher {
      override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
      
      }
      
      override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
      
      }
      
      override fun afterTextChanged(editable: Editable) {
        updatedCarbohydrates = Integer.parseInt(editable.toString())
      }
    })
    
    edit_entry_insulin_units.addTextChangedListener(object : TextWatcher {
      override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
      
      }
      
      override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
      
      }
      
      override fun afterTextChanged(editable: Editable) {
        updatedInsulin = java.lang.Double.parseDouble(editable.toString())
      }
    })
    
    
    
    delete_entry.setOnClickListener {
      val builder: AlertDialog.Builder = AlertDialog.Builder(this)
      builder.setTitle(R.string.delete_single_entry)
            .setMessage(R.string.delete_single_entry_message)
            .setPositiveButton(android.R.string.yes) { _, _ ->
              databaseManager.deleteEntry(old)
              Toast.makeText(this, R.string.entry_deleted, Toast.LENGTH_SHORT).show()
              startActivity(Intent(applicationContext, MainActivity::class.java))
            }
            .setNegativeButton(android.R.string.no) { dialog, _ ->
              dialog.dismiss()
            }
            .setIcon(R.drawable.ic_trash)
            .show()
    }
    update_fab.setOnClickListener { updateEntry() }
  }
  
  private fun getOriginalValues() {
    originalStatus = old.status
    originalCalendar.time = old.date
    originalDate = originalCalendar.time
    originalBloodGlucose = old.bloodGlucose
    originalCarbohydrates = old.carbohydrates
    originalInsulin = old.insulin
  }
  
  private fun setHints() {
    edit_status_selector.setSelection(originalStatus)
    edit_entry_date.hint = dateFormatter.formatDate(originalDate)
    edit_entry_time.hint = dateFormatter.twelveHourFormat(originalDate)
    edit_entry_blood_glucose_level.hint = originalBloodGlucose.toString()
    if (originalCarbohydrates != 0)
      edit_entry_carbohydrates_amount.hint = originalCarbohydrates.toString()
    else
      edit_entry_carbohydrates_amount.setHint(R.string.dash)
    if (originalInsulin != 0.0)
      edit_entry_insulin_units.hint = originalInsulin.toString()
    else
      edit_entry_insulin_units.setHint(R.string.dash)
  }
  
  private fun updateEntry() {
    val dateToSave = Date()
    val itemToSave = Entry()
    itemToSave.status = updatedStatus
    when (wasDateChanged) {
      true -> {
        dateToSave.time = updatedCalendar.timeInMillis
        itemToSave.date = dateToSave
      }
      false -> {
        dateToSave.time = originalCalendar.timeInMillis
        itemToSave.date = dateToSave
      }
    }
    if (edit_entry_blood_glucose_level.text.isNotEmpty()) itemToSave.bloodGlucose = edit_entry_blood_glucose_level.text.toString().toInt() else itemToSave.bloodGlucose = originalBloodGlucose
    if (edit_entry_carbohydrates_amount.text.isNotEmpty()) itemToSave.carbohydrates = edit_entry_carbohydrates_amount.text.toString().toInt() else itemToSave.carbohydrates = originalCarbohydrates
    if (edit_entry_insulin_units.text.isNotEmpty()) itemToSave.insulin = edit_entry_insulin_units.text.toString().toDouble() else itemToSave.insulin = originalInsulin
    databaseManager.deleteEntry(old)
    databaseManager.copyToRealm(itemToSave)
    finish()
  }
  
  override fun onSupportNavigateUp(): Boolean {
    onBackPressed()
    return super.onSupportNavigateUp()
  }
  
  override fun onItemSelected(p0: AdapterView<*>?, p1: View?, position: Int, p3: Long) {
    edit_status_selector.setSelection(position)
    updatedStatus = position
  }
  
  override fun onNothingSelected(p0: AdapterView<*>?) {}
}